Sources/XMLCoder/Encoder/XMLEncoderImplementation.swift

```diff
@@ -73,47 +73,85 @@ class XMLEncoderImplementation: Encoder {
+                    elements: KeyedBox.Elements([choice.withShared { ($0.key, $0.element) }]),
+                    attributes: []
+                )
+                let container = XMLKeyedEncodingContainer<Key>(
+                    referencing: self,
+                    codingPath: codingPath,
+                    wrapping: storage.pushKeyedContainer(keyed)
+                )
+                return KeyedEncodingContainer(container)
+            default:
                 preconditionFailure(
                     """
                     Attempt to push new keyed encoding container when already previously encoded \
                     at this path.
                     """
                 )
             }
-
-            topContainer = container
         }
-
-        let container = XMLKeyedEncodingContainer<Key>(referencing: self, codingPath: codingPath, wrapping: topContainer)
-        return KeyedEncodingContainer(container)
     }
 
     public func choiceContainer<Key>(keyedBy _: Key.Type) -> KeyedEncodingContainer<Key> {
-        let topContainer: SharedBox<ChoiceBox>
         if canEncodeNewValue {
             // We haven't yet pushed a container at this level; do so here.
-            topContainer = storage.pushChoiceContainer()
+            let container = XMLChoiceEncodingContainer<Key>(
+                referencing: self,
+                codingPath: codingPath,
+                wrapping: storage.pushChoiceContainer()
+            )
+            return KeyedEncodingContainer(container)
         } else {
-            guard let container = storage.lastContainer as? SharedBox<ChoiceBox> else {
+            switch storage.lastContainer {
+            case let keyed as SharedBox<KeyedBox>:
+                let container = XMLKeyedEncodingContainer<Key>(
+                    referencing: self,
+                    codingPath: codingPath,
+                    wrapping: keyed
+                )
+                return KeyedEncodingContainer(container)
+            case let choice as SharedBox<ChoiceBox>:
+                _ = storage.popContainer()
+                let keyed = KeyedBox(
+                    elements: KeyedBox.Elements([choice.withShared { ($0.key, $0.element) }]),
+                    attributes: []
+                )
+                let container = XMLKeyedEncodingContainer<Key>(
+                    referencing: self,
+                    codingPath: codingPath,
+                    wrapping: storage.pushKeyedContainer(keyed)
+                )
+                return KeyedEncodingContainer(container)
+            default:
                 preconditionFailure(
                     """
-                    Attempt to push new (single element) keyed encoding container when already \
-                    previously encoded at this path.
+                    Attempt to push new keyed encoding container when already previously encoded \
+                    at this path.
                     """
                 )
```

Comment:
Overall this `switch` seems to be a duplicate of `switch` in `public func keyedContainer` modified above, could it be generalized somehow?

Reply:
Ah, good point!